File: 797._All_Paths_From_Source_to_Target/allSourceToTarget.js
const allSourceToTarget = (graph) => {
  const paths = [];
  const route = [];
  const dfs = (graph, source, destination, route, paths) => {
    if (source === destination) {
      paths.push([...route]);
      return;
    }
    for (const element of graph[source]) {
      route.push(element);
      dfs(graph, element, destination, route, paths);
      route.pop();
    }
  };
  dfs(graph, 0, graph.length - 1, route, paths);
  return paths;
};

console.log(allSourceToTarget([[4, 3, 1], [3, 2, 4], [3], [4], []]));

File: 797._All_Paths_From_Source_to_Target/allSourceToTarget.py
def allSourceToTarget(graph):
  paths = []
  route = [0]
  def dfs(graph, source, destination, route, paths):
    if source == destination:
      paths.append(route.copy())
      return
    for element in graph[source]:
      route.append(element)
      dfs(graph, element, destination, route, paths)
      route.remove(element)
  dfs(graph, 0, len(graph) - 1, route, paths)
  return paths

print(allSourceToTarget([[4, 3, 1], [3, 2, 4], [3], [4], []]))
 


